include/wlr/types/wlr_pointer_constraints_v1.h

```diff
@@ -20,43 +24,91 @@ struct wlr_pointer_constraint_v1_state {
 };
 
 struct wlr_pointer_constraint_v1 {
+	struct wlr_pointer_constraints_v1 *pointer_constraints;
+
 	struct wl_resource *resource;
-	struct wl_resource *surface;
-	struct wl_resource *pointer;
+	struct wlr_surface *surface;
+	struct wlr_seat_client *seat_client;
 	uint32_t lifetime;
 
+	bool active;
+	bool accepted;
+
 	struct wlr_pointer_constraint_v1_state current, pending;
 
-	struct {
-		struct wl_signal destroy;
-	} events;
+	enum wlr_pointer_constraint_v1_type type;
 
-	struct wl_list link;
+	struct wl_listener surface_commit;
+	struct wl_listener surface_destroy;
 
-	enum wlr_pointer_constraint_v1_type type;
+	struct wl_list link; // wlr_surface::pointer_constraints
 
 	void *data;
 };
 
 struct wlr_pointer_constraints_v1 {
+	const struct wlr_pointer_constraints_v1_interface *impl;
 	struct wl_global *wl_global;
-
-	struct {
-		struct wl_signal constrain;
-	} events;
-
 	void* data;
 };
 
-struct wlr_pointer_constraints_v1 *wlr_pointer_constraints_v1_create(struct wl_display *display);
+struct wlr_pointer_constraints_v1_interface {
```

Comment:
Seems more appropriate to use `wl_signal` here

Reply:
Yeah I could probably do that. It's an interface because I initially had non-void return values.